### DISCUSSION TITLE
How to add libnoise to CMake

### DISCUSSION THREAD

**Discussion by dannypike** [microsoft/vcpkg#21227] [Q&A]
I'm new to CMake, so hopefully this is a simple misunderstanding ...

When I try to add the `libnoise` package to my CMakeLists.txt:

```
find_package(noise CONFIG REQUIRED)
target_link_libraries(noisy PRIVATE noise)
```

I get the following error:

```
1> [CMake] CMake Error at D:/work/github/vcpkg/scripts/buildsystems/vcpkg.cmake:788 (_find_package):
1> [CMake]   Could not find a package configuration file provided by "LibNoise" with any
1> [CMake]   of the following names:
1> [CMake] 
1> [CMake]     LibNoiseConfig.cmake
1> [CMake]     libnoise-config.cmake
1> [CMake] 
1> [CMake]   Add the installation prefix of "LibNoise" to CMAKE_PREFIX_PATH or set
1> [CMake]   "LibNoise_DIR" to a directory containing one of the above files.  If
1> [CMake]   "LibNoise" provides a separate development package or SDK, be sure it has
1> [CMake]   been installed.
```

Yet, `libpng` works just fine using a similar syntax:

```
find_package(libpng CONFIG REQUIRED)
target_link_libraries(noisy PRIVATE png)
```

presumably because that port does include a file called `libpng-config.cmake` in the `share/libpng` folder.

I did find a script called `FindLibNoise.cmake` in the `share\libnoise` folder of my _vcpkg_ installation, but I couldn't work out how to tell CMake where to find that, either:

```
include(libnoise/FindLibNoise.cmake)
```

produces this:

```
1> [CMake] CMake Error at noisy/CMakeLists.txt:12 (include):
1> [CMake]   include could not find requested file:
1> [CMake] 
1> [CMake]     libnoise/FindLibNoise.cmake
1> [CMake] 
1> [CMake] 
1> [CMake] -- Configuring incomplete, errors occurred!
```

I can't see anything in the _vcpkg_ documentation that explains to me how I should tell CMake where to find this script file (if I even should be doing that?)

Edit: I've tried using `noise`, `libnoise` and `LibNoise` to identify the port. The target library for libnoise is actually `noise.lib` but the target library for `libpng` is `libpng.lib` and I'm not sure what _vcpkg_ is going to expect me to say in the `find_package` statement. All things that I have tried have failed.

**Comment by dg0yt** [comment]
Did you try `find_package(LibNoise)`? That's how `FindLibNoise.cmake` is meant to be used - if the file is in CMake's module search path.

**Comment by dannypike** [comment]
I did (viz. my edit at the bottom of the file) and that fails with the same error.

So, I assume that there is a problem with how the _vcpkg_ toolchain file was setting up the CMake module search path, because `CMAKE_MODULE_PATH` is empty:

```
message(STATUS "module path ='${CMAKE_MODULE_PATH}'")
```

produces:

```
1> [CMake] -- module path =''
```

Incidentally, this is why I just posted this [comment](https://github.com/microsoft/vcpkg/issues/21226#issuecomment-962415526). I was going to build a `CMAKE_MODULE_PATH` myself, but I can't see how to include the _vcpkg_ triplet in the path.

**Reply by dg0yt** [reply]
>> Did you try `find_package(LibNoise)`?
>
> I did (viz. my edit at the bottom of the file) and that fails with the same error.

But this is what you wrote:

> ~~~
> include(libnoise/FindLibNoise.cmake)
> ~~~

---

> I was going to build a `CMAKE_MODULE_PATH` myself, but I can't see how to include the _vcpkg_ triplet in the path.

You can try `${VCPKG_INSTALLED_DIR}/share/libnoise`.

**Reply by dannypike** [reply]
> You can try ${VCPKG_INSTALLED_DIR}/share/libnois

Thank you for the suggestion but, unfortunately, `VCPKG_INSTALLED_DIR` does not include the triplet:

```
message(STATUS "path ='${VCPKG_INSTALLED_DIR}/share/libnoise/FindLibNoise.cmake'")
include(${VCPKG_INSTALLED_DIR}/share/libnoise/FindLibNoise.cmake)
```

results in this:

```
1> [CMake] -- path ='D:/work/github/vcpkg/installed/share/libnoise/FindLibNoise.cmake'
1> [CMake] CMake Error at noisy/CMakeLists.txt:14 (include):
1> [CMake]   include could not find requested file:
1> [CMake] 
1> [CMake]     D:/work/github/vcpkg/installed/share/libnoise/FindLibNoise.cmake
1> [CMake] 
1> [CMake] 
1> [CMake] -- Configuring incomplete, errors occurred!
```

when the actual path is `D:/work/github/vcpkg/installed/x64-windows/share/libnoise/FindLibNoise.cmake`

Hence I was looking for a variable that gives me the triplet :)

**Reply by dg0yt** [reply]
You are right, the triplet is not included. You need another variable.
~~~
${VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/share/libnoise
~~~

**Reply by dannypike** [reply]
Thank you!
